src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs

```diff
@@ -1,38 +1,61 @@
     {
         private IUnconfiguredProjectVsServices _unconfiguredProjectVsServices;
+        private readonly ProjectProperties _properties;
+        private readonly IVsFrameworkParser _frameworkParser;
+        private const string _targetFrameworkProperty = "TargetFramework";
+        private const string _targetFrameworksProperty = "TargetFrameworks";
 
         [ImportingConstructor]
-        public TargetFrameworkMonikerValueProvider(IUnconfiguredProjectVsServices unconfiguredProjectVsServices)
+        public TargetFrameworkMonikerValueProvider(IUnconfiguredProjectVsServices unconfiguredProjectVsServices, ProjectProperties properties, IVsFrameworkParser frameworkParser)
         {
             _unconfiguredProjectVsServices = unconfiguredProjectVsServices;
+            _properties = properties;
+            _frameworkParser = frameworkParser;
         }
 
-        public override Task<string> OnSetPropertyValueAsync(string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string> dimensionalConditions = null)
+        public override async Task<string> OnSetPropertyValueAsync(string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string> dimensionalConditions = null)
         {
-            // CPS implements IVsHierarchy.SetProperty for the TFM property to call through the multi-targeting service and change the TFM.
-            // This causes the project to be reloaded after changing the values.
-            // Since the property providers are called under a write-lock, trying to reload the project on the same context fails saying it can't load the project
-            // if a lock is held. We are not going to write to the file under this lock (we return null from this method) and so we fork execution here to schedule
-            // a lambda on the UI thread and we don't pass the lock information from this context to the new one. 
-            _unconfiguredProjectVsServices.ThreadingService.Fork(() =>
+            var configuration = await _properties.GetConfigurationGeneralPropertiesAsync().ConfigureAwait(true);
+            var currentTargetFramework = (string)await configuration.TargetFramework.GetValueAsync().ConfigureAwait(true);
+            var currentTargetFrameworks = (string)await configuration.TargetFrameworks.GetValueAsync().ConfigureAwait(true);
+            if (!string.IsNullOrEmpty(currentTargetFrameworks))
             {
-                _unconfiguredProjectVsServices.VsHierarchy.SetProperty(HierarchyId.Root, (int)VsHierarchyPropID.TargetFrameworkMoniker, unevaluatedPropertyValue);
-                return System.Threading.Tasks.Task.CompletedTask;
-            }, options: ForkOptions.HideLocks | ForkOptions.StartOnMainThread);
-
-            return System.Threading.Tasks.Task.FromResult<string>(null);
+                throw new InvalidOperationException(VSResources.MultiTFEditNotSupported);
```

Comment:
Are set properties allowed to throw? Do we leak in the project lock in this case? What does this look like in the project properties? Which TFM is shown? Why isn't the combo disabled in this case?